### ISSUE TITLE
Warning when non-delegate actual value is used with DelayedConstraint

### ISSUE THREAD

**Issue by Dreamescaper** [nunit/nunit.analyzers#149]
Case 1:
```csharp
int i = 0;
int actualFunc() => i++;
Assert.That(actualFunc(), Is.EqualTo(5).After(1000, 10));
```

DelayedConstraint is allowed to be used with non-delegate values, but that makes no sense, since it will never change, and most likely is not intended.
I suggest to add analyzer for this case, with Warning severity.

Case 2:
```csharp
Assert.That("123 ".Trim, Is.EqualTo("123"));
```
Opposite case. Delegate actual argument is allowed for any Constraint, but it will simply invoke it. 
It might be a bit confusing though, so I suggest adding analyzer for such cases (with Info severity).
Exception: async delegates.

**Comment by stevenaw** [comment]
(Apologies in advance for any mistakes below, I'm typing this from my phone)

I like this idea in that it will help catch the cases where one forgets to invoke a delegate when they meant to when writing tests.

One thought I have is around reference types. Should the following situation issue a warning?

```csharp
var myTest = new MyRefType();
// or
// ref var myTest = ref new MyStruct();
Assert.That(myTest, Has.Property("Foo").EqualTo("Bar").After(1000, 100))
```

**Comment by Dreamescaper** [comment]
Frankly speaking, haven't thought about such constraints. 
It technically might be valid for any non-immutable type.
Not sure how to handle that.

**Comment by Dreamescaper** [comment]
I suppose it could be limited to built-in types only.

**Comment by manfred-brands** [comment]
It is a pity that in the new constraint syntax we loose any compiler type constraints to prevent these. Like #357.
I will see if I can augment #359 to check for more cases than Throws.

**Comment by manfred-brands** [comment]
Case 1 is addressed by #359 
Case 2 might not look right, but NUnit will evaluate an ActualValueDelegate before the assert.
